### ISSUE TITLE
Transpiler fails on serialization internally

### ISSUE THREAD

**Issue by adekusar-drl** [Qiskit/qiskit-terra#9199]
### Environment

- **Qiskit Terra version**: latest(https://github.com/Qiskit/qiskit-terra/commit/479bde3000edb1c60b33450243817b05cfc7b1d0)
- **Python version**: 3.8
- **Operating system**: MacOS, Any

### What is happening?

The transpiler fails on serialization when it tries to parallelize transpilation of a list of circuits. Internally, `dill` fails to serialize an instance of the `PassManager`.

Originally the problem came from Qiskit Machine Learning, where the unit tests fail and CI is blocked.

@manoelmarques FYI 

### How can we reproduce the issue?

Run the script:
```python
import numpy as np
from qiskit import QuantumCircuit
from qiskit.circuit.library import ZZFeatureMap, RealAmplitudes
from qiskit.providers.fake_provider import FakeToronto
from qiskit.transpiler import PassManagerConfig
from qiskit.transpiler.preset_passmanagers import level_2_pass_manager


num_qubits = 2
feature_map = ZZFeatureMap(num_qubits, reps=1)
ansatz = RealAmplitudes(num_qubits, reps=1)

qc = QuantumCircuit(num_qubits)
qc.append(feature_map, range(2))
qc.append(ansatz, range(2))

pm = level_2_pass_manager(PassManagerConfig.from_backend(FakeToronto()))

tqc = pm.run(qc)
tqc_bound = tqc.bind_parameters(np.zeros(qc.num_parameters))

# fine
pm.run(tqc_bound)
pm.run(tqc_bound)

# fails
pm.run([tqc_bound, tqc_bound])
```

### What should happen?

The code fails:
```
dev-qml/bin/python __custom_pass_manager3.py 
thread '<unnamed>' panicked at 'Failed to set_item on dict: PyErr { type: <class 'TypeError'>, value: TypeError("unhashable type: 'list'"), traceback: None }', .../.cargo/registry/src/github.com-1ecc6299db9ec823/pyo3-0.17.3/src/types/dict.rs:423:18
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Traceback (most recent call last):
  File "__custom_pass_manager3.py", line 27, in <module>
    pm.run([tqc_bound, tqc_bound])
  File "dev-qml/lib/python3.8/site-packages/qiskit/transpiler/passmanager.py", line 528, in run
    return super().run(circuits, output_name, callback)
  File "dev-qml/lib/python3.8/site-packages/qiskit/transpiler/passmanager.py", line 231, in run
    return self._run_several_circuits(circuits, output_name, callback)
  File "dev-qml/lib/python3.8/site-packages/qiskit/transpiler/passmanager.py", line 265, in _run_several_circuits
    PassManager._in_parallel, circuits, task_kwargs={"pm_dill": dill.dumps(self)}
  File "dev-qml/lib/python3.8/site-packages/dill/_dill.py", line 263, in dumps
    dump(obj, file, protocol, byref, fmode, recurse, **kwds)#, strictio)
  File "dev-qml/lib/python3.8/site-packages/dill/_dill.py", line 235, in dump
    Pickler(file, protocol, **_kwds).dump(obj)
  File "dev-qml/lib/python3.8/site-packages/dill/_dill.py", line 394, in dump
    StockPickler.dump(self, obj)
  File "dev-qml/lib/python3.8/pickle.py", line 487, in dump
    self.save(obj)
  File "dev-qml/lib/python3.8/site-packages/dill/_dill.py", line 388, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "dev-qml/lib/python3.8/pickle.py", line 603, in save
    self.save_reduce(obj=obj, *rv)
  File "dev-qml/lib/python3.8/pickle.py", line 717, in save_reduce
    save(state)
  File "dev-qml/lib/python3.8/site-packages/dill/_dill.py", line 388, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "dev-qml/lib/python3.8/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "dev-qml/lib/python3.8/site-packages/dill/_dill.py", line 1186, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "dev-qml/lib/python3.8/pickle.py", line 971, in save_dict
    self._batch_setitems(obj.items())
  File "dev-qml/lib/python3.8/pickle.py", line 997, in _batch_setitems
    save(v)
  File "dev-qml/lib/python3.8/site-packages/dill/_dill.py", line 388, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "dev-qml/lib/python3.8/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "dev-qml/lib/python3.8/pickle.py", line 931, in save_list
    self._batch_appends(obj)
  File "dev-qml/lib/python3.8/pickle.py", line 955, in _batch_appends
    save(x)
  File "dev-qml/lib/python3.8/site-packages/dill/_dill.py", line 388, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "dev-qml/lib/python3.8/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "dev-qml/lib/python3.8/site-packages/dill/_dill.py", line 1186, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "dev-qml/lib/python3.8/pickle.py", line 971, in save_dict
    self._batch_setitems(obj.items())
  File "dev-qml/lib/python3.8/pickle.py", line 997, in _batch_setitems
    save(v)
  File "dev-qml/lib/python3.8/site-packages/dill/_dill.py", line 388, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "dev-qml/lib/python3.8/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "dev-qml/lib/python3.8/pickle.py", line 931, in save_list
    self._batch_appends(obj)
  File "dev-qml/lib/python3.8/pickle.py", line 958, in _batch_appends
    save(tmp[0])
  File "dev-qml/lib/python3.8/site-packages/dill/_dill.py", line 388, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "dev-qml/lib/python3.8/pickle.py", line 603, in save
    self.save_reduce(obj=obj, *rv)
  File "dev-qml/lib/python3.8/pickle.py", line 717, in save_reduce
    save(state)
  File "dev-qml/lib/python3.8/site-packages/dill/_dill.py", line 388, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "dev-qml/lib/python3.8/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "dev-qml/lib/python3.8/site-packages/dill/_dill.py", line 1186, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "dev-qml/lib/python3.8/pickle.py", line 971, in save_dict
    self._batch_setitems(obj.items())
  File "dev-qml/lib/python3.8/pickle.py", line 997, in _batch_setitems
    save(v)
  File "dev-qml/lib/python3.8/site-packages/dill/_dill.py", line 388, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "dev-qml/lib/python3.8/pickle.py", line 578, in save
    rv = reduce(self.proto)
pyo3_runtime.PanicException: Failed to set_item on dict: PyErr { type: <class 'TypeError'>, value: TypeError("unhashable type: 'list'"), traceback: None }

Process finished with exit code 1
```

**Comment by jakelishman** [comment]
I think I can see the problem here - it's `ErrorMap`, an internal Rust class for the VF2 layout passes causing issues.  Internally in Rust it uses a `HashMap<[usize; 2], f64>` to store the data, and to pickle itself it it has `__getstate__` and `__setstate__` methods that expose this object to Python via PyO3 binding.  The problem is that the natural PyO3 deserialisation type of Rust's `[usize; 2]` is Python's `list[int]`, which can't be used as a dict key, causing a Rust panic.

We can fix this by overriding PyO3 to ensure it converts the map to `dict[tuple[int, int], float]` instead, and then converts that back to the correct type in Rust.

**Comment by jakelishman** [comment]
Should be fixed by #9200.
